[plsql] ParseException with cursor declared in anonymous block

<!-- NB: issues about the rule designer should be opened at https://github.com/pmd/pmd-designer/issues -->

**Affects PMD Version:** 6.17.0

**Rule:**

**Description:** I get ParseException when trying to process file with single script with cursor in anonymous block declaration.

```
lip 23, 2019 2:12:18 PM net.sourceforge.pmd.RulesetsFactoryUtils printRuleNamesInDebug
FINER: Loaded rule AvoidTabCharacter
lip 23, 2019 2:12:18 PM net.sourceforge.pmd.PMD getApplicableLanguages
FINE: Using PLSQL version: PLSQL
lip 23, 2019 2:12:18 PM net.sourceforge.pmd.PMD processFiles
WARNING: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/latest/pmd_userdocs_incremental_analysis.html
lip 23, 2019 2:12:18 PM net.sourceforge.pmd.RulesetsFactoryUtils printRuleNamesInDebug
FINER: Loaded rule AvoidTabCharacter
lip 23, 2019 2:12:18 PM net.sourceforge.pmd.processor.PmdRunnable call
FINE: Processing C:\Users\...\01_DB\00_Scripts\sols83_icom_aq.sql
lip 23, 2019 2:12:18 PM net.sourceforge.pmd.processor.PmdRunnable addError
FINE: Error while processing file: C:\Users\...\01_DB\00_Scripts\sols83_icom_aq.sql
net.sourceforge.pmd.PMDException: Error while parsing C:\Users\...\01_DB\00_Scripts\sols83_icom_aq.sql
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:99)
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:51)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:78)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:24)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: net.sourceforge.pmd.lang.plsql.ast.ParseException: Encountered "" at line 1, column 1.
Was expecting one of:

        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.generateParseException(PLSQLParser.java)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.jj_consume_token(PLSQLParser.java)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Global(PLSQLParser.java:510)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Input(PLSQLParser.java:202)
        at net.sourceforge.pmd.lang.plsql.PLSQLParser.parse(PLSQLParser.java:55)
        at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:111)
        at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:175)
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:96)
        ... 9 more
```

**Code Sample demonstrating the issue:**

```
declare

  cursor c_queue_exists(cp_queue_name xla_reference_code.code%type) is
    select 1
    from   dba_queues dq
    where  dq.name = cp_queue_name;

  r_queue_exists c_queue_exists%rowtype;
begin
  null;
end;
/
```

**Running PMD through:** *[CLI]*

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -251,7 +251,7 @@ ASTInput Input(String sourcecode) : {}
 	 | LOOKAHEAD(6) Directory()
 	 | LOOKAHEAD(6) DatabaseLink()
 	 | LOOKAHEAD(6) Global()
-         | LOOKAHEAD(6) DDLCommand() //Ignore any other DDL Event
+     | LOOKAHEAD(6) DDLCommand() //Ignore any other DDL Event
 	 | LOOKAHEAD(2) SqlPlusCommand()
 	 | UpdateStatement()
 	 | DeleteStatement()
@@ -351,20 +351,18 @@ ASTGlobal Global()  :
 {
 }
 {
-
 	/*
 	  Remove triggers from global processing because their schema may be defined in the trigger code itself
 	  Still wrap the trigger in a fake package but make the package name dependent on the actual schema
 	  defaulting to the globalPackageName if it cannot be found
 	*/
-	(LOOKAHEAD ( ( Label() )* [<DECLARE> DeclarativeSection()] <BEGIN>) ( Label() )* Block() ";" | LOOKAHEAD (4) ProgramUnit()
+	(
+	LOOKAHEAD ( ( Label() )* ( <DECLARE> | <BEGIN> ) ) ( Label() )* Block() ";"
+	|
+	LOOKAHEAD (4) ProgramUnit()
 	)
-
-
       { return jjtThis ; }
 }
-
-
 
 ASTBlock Block()  :
 {

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/AnonymousBlockTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/AnonymousBlockTest.java
@@ -1,0 +1,32 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.plsql.ast;
+
+import java.nio.charset.StandardCharsets;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import net.sourceforge.pmd.lang.plsql.AbstractPLSQLParserTst;
+
+public class AnonymousBlockTest extends AbstractPLSQLParserTst {
+
+    @Test
+    public void parseCursorInsideProcAnonymousBlock() throws Exception {
+        String code = IOUtils.toString(this.getClass().getResourceAsStream("AnonymousBlock1.sql"),
+                StandardCharsets.UTF_8);
+        ASTInput input = parsePLSQL(code);
+        Assert.assertNotNull(input);
+    }
+
+    @Test
+    public void parseCursorInsideAnonymousBlock() throws Exception {
+        String code = IOUtils.toString(this.getClass().getResourceAsStream("AnonymousBlock2.sql"),
+                StandardCharsets.UTF_8);
+        ASTInput input = parsePLSQL(code);
+        Assert.assertNotNull(input);
+    }
+}

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/AnonymousBlock1.sql
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/AnonymousBlock1.sql
@@ -1,0 +1,21 @@
+declare
+  gw_acl_name   varchar2(100);
+
+  procedure create_acl(p_acl_name in varchar2) is
+    cursor c_acl is
+      select dna.acl
+      from   dba_network_acls dna
+      where  acl like '%' || p_acl_name;
+    w_acl_name varchar2(100);
+  begin
+    open c_acl;
+    fetch c_acl
+      into w_acl_name;
+    close c_acl;
+
+  end;
+
+begin
+  null;
+end;
+/

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/AnonymousBlock2.sql
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/AnonymousBlock2.sql
@@ -1,0 +1,12 @@
+declare
+
+  cursor c_queue_exists(cp_queue_name xla_reference_code.code%type) is
+    select 1
+    from   dba_queues dq
+    where  dq.name = cp_queue_name;
+
+  r_queue_exists c_queue_exists%rowtype;
+begin
+  null;
+end;
+/

